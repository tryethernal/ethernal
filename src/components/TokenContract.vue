<template>
    <v-container v-if="loadingContract" fluid>
        <v-skeleton-loader
            type="text"
            max-width="600"
            class="mb-4"
        />
        <v-divider class="my-4"></v-divider>
        
        <!-- Skeleton for token info cards -->
        <v-row class="mb-6">
            <v-col cols="12" lg="4">
                <v-skeleton-loader type="card" height="200" />
            </v-col>
            <v-col cols="12" lg="4">
                <v-skeleton-loader type="card" height="200" />
            </v-col>
            <v-col cols="12" lg="4">
                <v-skeleton-loader type="card" height="200" />
            </v-col>
        </v-row>
    </v-container>
    <v-container v-else-if="notAContract" fluid>
        <v-row justify="center" align="center" style="min-height: 400px">
            <v-col cols="12" sm="8" md="6" lg="4" class="text-center">
                <v-icon size="100" color="primary" class="mb-4" style="opacity: 0.25">mdi-file</v-icon>
                <div class="text-h6 mb-2">No Contract Found</div>
                <div class="text-body-1 text-medium-emphasis">There doesn't seem to be a contract at this address.</div>
            </v-col>
        </v-row>
    </v-container>
    <ERC20Contract
        v-else-if="isERC20"
        :address="address"
        :contract="contract"
        :loading-contract="loadingContract"
        :key="`erc20-${address}`"
    />
    <ERC721Collection
        v-else-if="isERC721"
        :address="address"
        :contract="contract"
        :loading-contract="loadingContract"
        :key="`erc721-${address}`"
    />
    <v-container v-else fluid>
        <v-row justify="center" align="center" style="min-height: 400px">
            <v-col cols="12" sm="8" md="6" lg="4" class="text-center">
                <v-icon size="100" color="primary" class="mb-4" style="opacity: 0.25">mdi-help-circle</v-icon>
                <div class="text-h6 mb-2">Unknown Contract Type</div>
                <div class="text-body-1 text-medium-emphasis">This contract doesn't appear to be an ERC20 or ERC721 token.</div>
            </v-col>
        </v-row>
    </v-container>
</template>

<script setup>
import { ref, computed, watch, inject } from 'vue';
import ERC20Contract from './ERC20Contract.vue';
import ERC721Collection from './ERC721Collection.vue';

// Props
const props = defineProps({
    address: {
        type: String,
        required: true
    }
});

// Inject server instance
const $server = inject('$server');

// Reactive state
const loadingContract = ref(true);
const contract = ref({});
const notAContract = ref(false);

// Computed properties
const isERC20 = computed(() => {
    return contract.value?.patterns?.includes('erc20');
});

const isERC721 = computed(() => {
    return contract.value?.patterns?.includes('erc721') || 
           contract.value?.patterns?.includes('erc1155');
});

// Load contract data
const loadContract = async (address) => {
    try {
        loadingContract.value = true;
        notAContract.value = false;
        contract.value = {};

        const { data } = await $server.getContract(address);
        if (data) {
            contract.value = data;
        } else {
            notAContract.value = true;
        }
    } catch (error) {
        console.error('Error loading contract:', error);
        notAContract.value = true;
    } finally {
        loadingContract.value = false;
    }
};

// Watchers
watch(() => props.address, (newAddress) => {
    if (newAddress) {
        loadContract(newAddress);
    }
}, { immediate: true });
</script> 