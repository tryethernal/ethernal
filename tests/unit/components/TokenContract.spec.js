import TokenContract from '@/components/TokenContract.vue';

const stubs = ['ERC20Contract', 'ERC721Collection'];

describe('TokenContract.vue', () => {
    it('Should show loading state', async () => {
        vi.spyOn(server, 'getContract').mockImplementationOnce(() => new Promise(() => {}));

        const wrapper = mount(TokenContract, {
            props: {
                address: '0x123'
            },
            global: {
                stubs
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show not a contract message', async () => {
        vi.spyOn(server, 'getContract').mockResolvedValueOnce({ data: null });

        const wrapper = mount(TokenContract, {
            props: {
                address: '0x123'
            },
            global: {
                stubs
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show ERC20 contract component', async () => {
        vi.spyOn(server, 'getContract').mockResolvedValueOnce({
            data: {
                patterns: ['erc20']
            }
        });

        const wrapper = mount(TokenContract, {
            props: {
                address: '0x123'
            },
            global: {
                stubs
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show ERC721 contract component', async () => {
        vi.spyOn(server, 'getContract').mockResolvedValueOnce({
            data: {
                patterns: ['erc721']
            }
        });

        const wrapper = mount(TokenContract, {
            props: {
                address: '0x123'
            },
            global: {
                stubs
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show ERC1155 contract as ERC721 component', async () => {
        vi.spyOn(server, 'getContract').mockResolvedValueOnce({
            data: {
                patterns: ['erc1155']
            }
        });

        const wrapper = mount(TokenContract, {
            props: {
                address: '0x123'
            },
            global: {
                stubs
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show unknown contract type message', async () => {
        vi.spyOn(server, 'getContract').mockResolvedValueOnce({
            data: {
                patterns: ['unknown']
            }
        });

        const wrapper = mount(TokenContract, {
            props: {
                address: '0x123'
            },
            global: {
                stubs
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show not a contract message when getContract throws error', async () => {
        vi.spyOn(server, 'getContract').mockRejectedValueOnce(new Error('Failed to load contract'));

        const wrapper = mount(TokenContract, {
            props: {
                address: '0x123'
            },
            global: {
                stubs
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });
}); 