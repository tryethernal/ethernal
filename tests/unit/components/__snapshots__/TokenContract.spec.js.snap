// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TokenContract.vue > Should show ERC20 contract component 1`] = `"<e-r-c20-contract-stub address="0x123" contract="[object Object]" loadingcontract="false"></e-r-c20-contract-stub>"`;

exports[`TokenContract.vue > Should show ERC721 contract component 1`] = `"<e-r-c721-collection-stub address="0x123" contract="[object Object]" loadingcontract="false"></e-r-c721-collection-stub>"`;

exports[`TokenContract.vue > Should show ERC1155 contract as ERC721 component 1`] = `"<e-r-c721-collection-stub address="0x123" contract="[object Object]" loadingcontract="false"></e-r-c721-collection-stub>"`;

exports[`TokenContract.vue > Should show loading state 1`] = `
"<div class="v-container v-container--fluid v-locale--is-ltr">
  <div class="v-skeleton-loader v-theme--light mb-4" style="max-width: 600px;" arialive="polite" arialabel="Loading..." role="alert">
    <div class="v-skeleton-loader__bone v-skeleton-loader__text"></div>
  </div>
  <hr class="v-divider v-theme--light my-4" aria-orientation="horizontal" role="separator"><!-- Skeleton for token info cards -->
  <div class="v-row mb-6">
    <div class="v-col-lg-4 v-col-12">
      <div class="v-skeleton-loader v-theme--light" style="height: 200px;" arialive="polite" arialabel="Loading..." role="alert">
        <div class="v-skeleton-loader__bone v-skeleton-loader__card">
          <div class="v-skeleton-loader__bone v-skeleton-loader__image"></div>
          <div class="v-skeleton-loader__bone v-skeleton-loader__heading"></div>
        </div>
      </div>
    </div>
    <div class="v-col-lg-4 v-col-12">
      <div class="v-skeleton-loader v-theme--light" style="height: 200px;" arialive="polite" arialabel="Loading..." role="alert">
        <div class="v-skeleton-loader__bone v-skeleton-loader__card">
          <div class="v-skeleton-loader__bone v-skeleton-loader__image"></div>
          <div class="v-skeleton-loader__bone v-skeleton-loader__heading"></div>
        </div>
      </div>
    </div>
    <div class="v-col-lg-4 v-col-12">
      <div class="v-skeleton-loader v-theme--light" style="height: 200px;" arialive="polite" arialabel="Loading..." role="alert">
        <div class="v-skeleton-loader__bone v-skeleton-loader__card">
          <div class="v-skeleton-loader__bone v-skeleton-loader__image"></div>
          <div class="v-skeleton-loader__bone v-skeleton-loader__heading"></div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`TokenContract.vue > Should show not a contract message 1`] = `
"<div class="v-container v-container--fluid v-locale--is-ltr">
  <div class="v-row align-center justify-center" style="min-height: 400px;">
    <div class="v-col-sm-8 v-col-md-6 v-col-lg-4 v-col-12 text-center"><i class="mdi-file mdi v-icon notranslate v-theme--light text-primary mb-4" style="font-size: 100px; height: 100px; width: 100px; opacity: 0.25;" aria-hidden="true"></i>
      <div class="text-h6 mb-2">No Contract Found</div>
      <div class="text-body-1 text-medium-emphasis">There doesn't seem to be a contract at this address.</div>
    </div>
  </div>
</div>"
`;

exports[`TokenContract.vue > Should show not a contract message when getContract throws error 1`] = `
"<div class="v-container v-container--fluid v-locale--is-ltr">
  <div class="v-row align-center justify-center" style="min-height: 400px;">
    <div class="v-col-sm-8 v-col-md-6 v-col-lg-4 v-col-12 text-center"><i class="mdi-file mdi v-icon notranslate v-theme--light text-primary mb-4" style="font-size: 100px; height: 100px; width: 100px; opacity: 0.25;" aria-hidden="true"></i>
      <div class="text-h6 mb-2">No Contract Found</div>
      <div class="text-body-1 text-medium-emphasis">There doesn't seem to be a contract at this address.</div>
    </div>
  </div>
</div>"
`;

exports[`TokenContract.vue > Should show unknown contract type message 1`] = `
"<div class="v-container v-container--fluid v-locale--is-ltr">
  <div class="v-row align-center justify-center" style="min-height: 400px;">
    <div class="v-col-sm-8 v-col-md-6 v-col-lg-4 v-col-12 text-center"><i class="mdi-help-circle mdi v-icon notranslate v-theme--light text-primary mb-4" style="font-size: 100px; height: 100px; width: 100px; opacity: 0.25;" aria-hidden="true"></i>
      <div class="text-h6 mb-2">Unknown Contract Type</div>
      <div class="text-body-1 text-medium-emphasis">This contract doesn't appear to be an ERC20 or ERC721 token.</div>
    </div>
  </div>
</div>"
`;
